src/gpacmp4/mp4.c

```diff
@@ -304,20 +304,196 @@ unsigned char * ccdp_find_data(unsigned char * ccdp_atom_content, unsigned int l
+				cb_708++;
+			}
+		}
+		else //subtype == GF_ISOM_SUBTYPE_C608
+		{
+			if (is_ccdp) {
+				mprint("Your video file seems to be an interesting sample for us\n");
+				mprint("We haven't met c608 subtitle in a \'ccdp\' atom before\n");
+				mprint("Please, report\n");
+				return -1;
+			}
+
+			int ret = 0;
+			int len = atom_length - 8;
+			data += 4;
+			char *tdata = data;
+			do {
+				// Process each pair independently so we can adjust timing
+				ret = process608((unsigned char *)tdata, len > 2 ? 2 : len, dec_ctx, dec_sub);
+				len -= ret;
+				tdata += ret;
+				cb_field1++;
+				if (dec_sub->got_output) {
+					*mp4_ret = 1;
+					encode_sub(enc_ctx, dec_sub);
+					dec_sub->got_output = 0;
+				}
+			} while (len > 0);
+		}
+	}
+	return atom_length;
+}
+
+// Rrocess tx3g type atom
+// Return the length of the atom
+// Return -1 if unrecoverable error happened or process of the atom is finished.
+// In this case, the sample will be skipped.
+// Argument `encode_last_only` is used to print the last subtitle
+static int process_tx3g(struct lib_ccx_ctx *ctx, struct encoder_ctx *enc_ctx,
+	struct lib_cc_decode *dec_ctx, struct cc_subtitle *dec_sub, int *mp4_ret,
+	char *data, size_t data_length, int encode_last_only) {
+	// tx3g data format:
+	// See https://developer.apple.com/library/content/documentation/QuickTime/QTFF/QTFFChap3/qtff3.html (Section Subtitle Sample Data)
+	// See http://www.etsi.org/deliver/etsi_ts/126200_126299/126245/14.00.00_60/ts_126245v140000p.pdf (Section 5.17)
+	// Basically it's 16-bit length + UTF-8/UTF-16 text + optional extension (ignored)
+	// TODO: support extension & style
+
+	static struct cc_subtitle last_sub;
+	static int has_previous_sub = 0;
+
+	// Always encode the previous subtitle, no matter the current one is valid or not
+	if (has_previous_sub) {
+		dec_sub->end_time = dec_ctx->timing->fts_now;
+		encode_sub(enc_ctx, dec_sub); // encode the previous subtitle
+		has_previous_sub = 0;
+	}
+	if (encode_last_only) return;
+
+	unsigned int atom_length = RB16(data);
+	data += 2;
+	
+	if (atom_length > data_length) {
+		mprint("Invalid atom length. Atom length: %u, should be: %u\n", atom_length, data_length);
+		return -1;
+	}
+	if (atom_length == 0) {
+		return -1;
+	}
+#ifdef MP4_DEBUG
+	dump(256, (unsigned char *)data, atom_length, 0, 1);
+#endif
+
+	// Start encode the current subtitle
+	// But they won't be written to file until we get the next subtitle
+	// So that we can know its end time
+	dec_sub->type = CC_TEXT;
+	dec_sub->start_time = dec_ctx->timing->fts_now;
+	if (dec_sub->data != NULL) free(dec_sub->data);
+	dec_sub->data = malloc(atom_length + 1);
+	memcpy(dec_sub->data, data, atom_length);
+	*((char*)dec_sub->data + atom_length) = '\0';
+
+	*mp4_ret = 1;
+	has_previous_sub = 1;
+
+	return -1;
+}
+
+#define MEDIA_TYPE(type, subtype) (((u64)(type)<<32)+(subtype))
+
+void switch_output_file(struct lib_ccx_ctx *ctx, struct encoder_ctx *enc_ctx, int track_id) {
```

Comment:
Let's call the function mp4_switch_output_file()
OR move to the file that contains all the file related stuff